src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs

```diff
@@ -0,0 +1,90 @@
+// <copyright file="LazyInstrumentationLoader.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using OpenTelemetry.AutoInstrumentation.Configuration;
+using OpenTelemetry.AutoInstrumentation.Logging;
+
+namespace OpenTelemetry.AutoInstrumentation.Loading
+{
+    internal class LazyInstrumentationLoader : IDisposable
+    {
+        private static readonly ILogger Logger = OtelLogging.GetLogger();
+
+        private readonly List<AssemblyLoadDetector> _assemblyLoadDetectors = new();
+        private readonly ConcurrentBag<object> _instrumentations = new();
+        private readonly TracerSettings _tracerSettings;
+
+        public LazyInstrumentationLoader(TracerSettings tracerSettings)
+        {
+            _tracerSettings = tracerSettings;
+
+            Subscribe(TracerInstrumentation.AspNet, () => new AspNetCoreDetector());
+        }
+
+        public void Dispose()
+        {
+            while (_instrumentations.TryTake(out object instrumentation))
+            {
+                if (instrumentation is IDisposable disposableInstrumentation)
+                {
+                    disposableInstrumentation.Dispose();
+                }
+            }
+        }
+
+        private void Subscribe(TracerInstrumentation instrumentation, Func<AssemblyLoadDetector> builder)
+        {
+            if (_tracerSettings.EnabledInstrumentations.Contains(instrumentation))
```

Comment:
I suggest moving this `if` to the caller. `_tracerSettings` could be removed and this class would be easier to test and also easier to reuse for metrics instrumentations.

Reply:
This would then generate multiple ifs instead of one.

Reply:
Check if it won't be better if `Subscribe` calls would be moved to `Instrumentation`. Otherwise, I think, it would be hard to unit test the assembly loading + instrumentation loading code.